### PR TITLE
relooper implementation

### PR DESCRIPTION
**Pull request by shishkin-pavel**

### CODE REVIEW

File: bin/evm2near/src/analyze.rs [hunk 1]
```diff
@@ -1,152 +1,129 @@
...
-    collections::{BTreeMap, BTreeSet},
+    collections::HashMap,
+    fmt::{Debug, Display},
     ops::Range,
 };
 
-#[derive(Clone, Debug, Hash, PartialEq, Eq)]
-pub struct CFGProgram(pub BTreeMap<Label, Block>);
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Offs(pub usize);
```
Comment (birchmd) [inline bin/evm2near/src/analyze.rs:14]: A doc comment describing what this newtype is for would be appreciated. A more descriptive name might be appropriate too.

One heuristic to use when thinking about how descriptive of a name to give is to make the level of description proportional to the size of the scope where that name can be used. For example it is fine to have `x` as a variable name in a one-line closure because the scope is so tiny, but on the hand a `pub struct` like this has scope equal to the entire program so a good name is warranted.

File: bin/evm2near/src/analyze.rs [hunk 2]
```diff
@@ -1,152 +1,129 @@
...
 
-pub type Label = usize;
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Idx(pub usize);
```
Comment (birchmd) [inline bin/evm2near/src/analyze.rs:26]: Same comment as above.

File: bin/evm2near/src/analyze.rs [hunk 3]
```diff
@@ -1,152 +1,129 @@
...
+impl Display for Idx {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
     }
 }
 
-impl Default for Block {
-    fn default() -> Self {
-        Self::new()
+struct BlockStart(Offs, Idx, bool);
```
Comment (birchmd) [inline bin/evm2near/src/analyze.rs]: It would probably make the code more readable for this to have named fields as opposed to being a tuple struct. The `bool` in particular is especially egregious because it is impossible to know what it represents from the type alone.

File: bin/evm2near/src/analyze.rs [hunk 4]
```diff
@@ -1,152 +1,129 @@
...
-    #[allow(dead_code)]
-    pub fn dump(&self) {
-        for block in self.0.values() {
-            println!("{:?}", block) // DEBUG
-        }
-    }
+#[derive(Debug)]
+pub struct BasicCfg {
+    pub cfg: Cfg<Offs>,
+    pub node_info: HashMap<Offs, (bool, bool)>,
```
Comment (birchmd) [inline bin/evm2near/src/analyze.rs]: I know you have a comment saying what the `bool`s mean a few lines down, but I think it is nice when someone looking at the code can figure out what things are for without needing to read the implementation of methods. I would either move the comment up here as a doc comment, or even better you could make a little `NodeInfo` struct with named fields for the `bool`s. Then you also don't need to remember which is which when matching on the tuple later in the code.

File: bin/evm2near/src/compile.rs [hunk 5]
```diff
@@ -111,24 +140,33 @@ impl Compiler {
...
+            full_path.push(path.into());
+
+            std::fs::write(full_path, contents()).expect("fs error while writing debug file");
+        }
+    }
+
     /// Emit an empty `_start` function to make all WebAssembly runtimes happy.
     fn emit_wasm_start(self: &mut Compiler) {
-        _ = self.emit_function(Some("_start".to_string()), vec![]);
+        _ = self.emit_function(Some("_start".to_string()), vec![Instruction::Unreachable]);
```
Comment (birchmd) [inline bin/evm2near/src/compile.rs]: Does this really work? I thought the start function was always called in the Wasm runtimes so I would expect that execution will always terminate immediately by hitting the unreachable opcode here.

File: bin/evm2near/src/compile.rs [hunk 6]
```diff
@@ -232,253 +270,269 @@ impl Compiler {
...
-    /// Compiles the program's control-flow graph.
-    fn compile_cfg(self: &mut Compiler, input_cfg: &CFGProgram, input_program: &Program) {
-        assert_ne!(self.evm_start_function, 0); // filled in during emit_start()
-        assert_eq!(self.evm_exec_function, 0); // filled in below
+    fn relooped_cfg(&self, basic_cfg: &BasicCfg) -> ReSeq<SLabel<CaterpillarLabel<EvmBlock>>> {
+        let cfg = basic_cfg.cfg.map_label(|label| {
+            let code_range = basic_cfg
+                .code_ranges
+                .get(label)
+                .unwrap_or_else(|| panic!("no code ranges for {}", *label));
```
Comment (birchmd) [inline bin/evm2near/src/compile.rs]: The `*` is probably not needed here.

File: bin/evm2near/src/compile.rs [hunk 7]
```diff
@@ -232,253 +270,269 @@ impl Compiler {
...
+        program: &Program,
+        cfg_part: &ReSeq<SLabel<CaterpillarLabel<EvmBlock>>>,
+        res: &mut Vec<Instruction>,
+        wasm_idx2evm_idx: &mut HashMap<Idx, Idx>,
+    ) {
+        for block in cfg_part.0.iter() {
+            match block {
+                ReBlock::Block(inner_seq) => {
+                    res.push(Instruction::Block(BlockType::NoResult));
+                    self.unfold_cfg(program, inner_seq, res, wasm_idx2evm_idx);
```
Comment (birchmd) [inline bin/evm2near/src/compile.rs:320]: Out of curiosity, how deeply nested do we expect these structures to be? Is this recursive call in danger of causing a stack overflow for some complex contracts?

File: bin/evm2near/src/compile.rs [hunk 8]
```diff
@@ -232,253 +270,269 @@ impl Compiler {
...
+                                            ]);
+                                        }
+                                        if op.is_push() {
+                                            let operands = encode_push(op);
+                                            res.extend(operands);
+                                        }
+                                        let call = self.compile_operator(op);
+                                        res.push(call);
+                                        if op == &Opcode::RETURN {
+                                            //TODO idk
```
Comment (birchmd) [inline bin/evm2near/src/compile.rs]: Remaining todo

File: bin/evm2near/src/compile.rs [hunk 9]
```diff
@@ -232,253 +270,269 @@ impl Compiler {
...
+                                            res.push(Instruction::Return);
+                                        }
+                                        curr_idx += 1;
+                                        evm_offset += op.size();
+                                    }
+                                    [] => {}
```
Comment (birchmd) [inline bin/evm2near/src/compile.rs]: Could leave the `unreachable` here? Or is this case possible now?

File: bin/evm2near/src/compile.rs [hunk 10]
```diff
@@ -232,253 +270,269 @@ impl Compiler {
...
+        self.debug("relooped.dot", || {
+            format!("digraph {{{}}}", relooped_cfg.to_dot())
+        });
+
+        let mut wasm: Vec<Instruction> = Default::default();
+        let mut wasm_idx2evm_idx = Default::default();
+        self.unfold_cfg(program, &relooped_cfg, &mut wasm, &mut wasm_idx2evm_idx);
+        wasm.push(Instruction::End);
+
+        self.evm_wasm_dot_debug(program, &basic_cfg, &relooped_cfg, &wasm, &wasm_idx2evm_idx);
```
Comment (birchmd) [inline bin/evm2near/src/compile.rs]: This should only be called if `config.denug_path.is_some()` right?

File: lib/evmlib/src/api.rs [hunk 11]
```diff
@@ -142,8 +137,19 @@ pub unsafe fn _evm_pop_u32() -> u32 {
...
+
 #[no_mangle]
 pub unsafe fn _evm_set_pc(pc: u32) {
     #[cfg(feature = "pc")]
     EVM.program_counter = pc;
 }
+
+#[no_mangle]
+pub unsafe fn _evm_burn_gas(gas: u32) {
+    // TODO gas value should be u64
```
Comment (birchmd) [inline lib/evmlib/src/api.rs:153]: Remaining todo

File: lib/evmlib/src/ops.rs [hunk 12]
```diff
@@ -659,90 +655,96 @@ pub unsafe fn pop() {
...
-    trace!("JUMP");
     // We only do JUMP gas cost accounting here, the actual branch is
     // synthesized by the compiler.
     EVM.burn_gas(8);
+    trace!("JUMP");
 }
 
 #[no_mangle]
-pub unsafe fn jumpi() -> u32 {
+pub unsafe fn jumpi() {
```
Comment (birchmd) [inline lib/evmlib/src/ops.rs]: Is this function used any more? Looked to me like all the logic (including the EVM gas burning) has been moved to the Wasm codegen directly.

File: lib/relooper/BACKLOG.md [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# Here will be a list of potential improvements.
+1) Rewrite domination tree building algorithm.
+     There is an algorithm with better time complexety but much harder to implement.
+2) Find optimal order of split and merge in reducing algorithm
+     Different orders generate different graphs and different graphs provide different performance (both gas and time)
+     of output programs
+3) `TLabel` or `&TLabel` or should we rely on `TLabel` only (giving user possibility to choose between structs/references)?
```
Comment (birchmd) [inline lib/relooper/BACKLOG.md]: `TLabel` could be a reference itself even if it is not explicitly declared as such. References are types in Rust just like structs so either could be used to fill in a type parameter.

For example, look at the definition of `Option` in the standard library, it just uses `T` as the type parameter, and yet I can construct an `Option` that holds a reference.

All this to say, this TODO may not be necessary.

File: lib/relooper/README.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+
+Relooper generates structured control flow graph (classical control flow statements like ``if``, ``for``, etc.)
+from unstructured (``goto`` control flow statements).
+
+This implementation is based on [paper](https://dl.acm.org/doi/pdf/10.1145/3547621)
+
+To see docs write next commands
+```
+git clone https://github.com/shishkin-pavel/relooper/
```
Comment (birchmd) [inline lib/relooper/README.md]: Why do they need to clone your repo? Could the docs not be viewed here?

File: lib/relooper/src/graph/cfg/mod.rs [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+        self.out_edges
+            .get(label)
+            .into_iter()
+            .flat_map(|edge| edge.to_vec())
+            .collect()
+    }
+
+    pub fn add_edge(&mut self, from: T, edge: CfgEdge<T>) {
+        let nodes = self.nodes();
+        // todo unncecessary collect
```
Comment (birchmd) [inline lib/relooper/src/graph/cfg/mod.rs]: Remaining todo. And this one is easy to fix, literally just remove the `collect` on line 97.

File: lib/relooper/src/graph/cfg/mod.rs [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        for n in edge
+            .to_vec()
+            .into_iter()
+            .filter(|n| !nodes.contains(n))
+            .collect::<Vec<_>>()
+        {
+            // todo clone
```
Comment (birchmd) [inline lib/relooper/src/graph/cfg/mod.rs]: Remaining todo

File: lib/relooper/src/graph/cfg/mod.rs [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    pub fn from_edges(entry: TLabel, edges: &HashMap<TLabel, CfgEdge<TLabel>>) -> Self {
+        let mut cfg = Cfg::new(entry);
+        for (&from, &edge) in edges.iter() {
+            cfg.add_edge(from, edge);
+        }
+
+        cfg
+    }
+
+    #[deprecated]
```
Comment (birchmd) [inline lib/relooper/src/graph/cfg/mod.rs]: I don't think we should introduce new code that is already deprecated. If we are not using it any more then simply remove it. It's not like we have other people who might be using this library because we just wrote it.

File: lib/relooper/src/graph/cfg/mod.rs [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+        in_edges
+    }
+
+    fn reachable_nodes(&self) -> HashSet<&TLabel> {
+        Bfs::start_from(&self.entry, |label| self.children(label)).collect()
+    }
+
+    pub fn strip_unreachable(&mut self) {
+        let nodes: HashSet<TLabel> = self.nodes().into_iter().copied().collect(); // TODO get rid of copies
```
Comment (birchmd) [inline lib/relooper/src/graph/cfg/mod.rs]: Remaining todo. This one should also be easy to fix. Remove the call to `copied` and change the type to `HashSet<&TLabel>` should work.

File: lib/relooper/src/graph/mod.rs [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+            cfg,
+            node_ordering,
+            domination,
+            merge_nodes,
+            loop_nodes,
+            if_nodes,
+        }
+    }
+
+    //TODO move to enrichments' DomTree constructor
```
Comment (birchmd) [inline lib/relooper/src/graph/mod.rs]: Remaining todo. If this is a significant amount of work then maybe file as a GitHub issue instead of fixing as part of this PR.

File: lib/relooper/src/graph/mod.rs [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        node_ordering: &NodeOrdering<TLabel>,
+        begin: TLabel,
+    ) -> HashMap<TLabel, TLabel> /* map points from node id to id of its dominator */ {
+        let mut result = HashMap::<TLabel, TLabel>::new();
+        let mut bfs = VecDeque::<TLabel>::new();
+        let mut visited = HashSet::<TLabel>::new();
+        for &n in node_ordering.sequence() {
+            result.insert(n, begin);
+        }
+        bfs.push_back(begin); // should be next. upd: i dont think so
```
Comment (birchmd) [inline lib/relooper/src/graph/mod.rs]: Confusing comment. Remember that comments are not just for other people, they are also for our future selves. Be kind to your future self and explain why it should not be "next".

File: lib/relooper/src/graph/supergraph.rs [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+            .get(&node.head)
+            .into_iter()
+            .flatten()
+            .map(|l| self.label_location.get(l).unwrap())
+            .map(|snode_label| self.nodes.get(snode_label).unwrap())
+            .collect();
+        // if given node have internal edges ending in its head, it will be seen as incoming supernode, which isn't useful
+        incoming.remove(node);
+
+        // TODO hate there is no pattern-match adapters for simple collections, or is there?
```
Comment (birchmd) [inline lib/relooper/src/graph/supergraph.rs]: Yeah, I don't think there is structural mapping on `BTreeSet`. And to be fair I think it would be a tricky thing to add to the language. Rust pattern matching is generally on the literal syntactical structure of the type, not the logical thing it represents, and in the case of `BTreeSet` you wouldn't want the developer to have to know the internal details of the `BTree` to use the pattern match.

File: lib/relooper/src/graph/supergraph.rs [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+
+            for (split_node, split_for) in splits.iter() {
+                split_len
+                    .entry(split_for.len())
+                    .or_default()
+                    .push(split_node);
+            }
+
+            if let Some((_, biggest_splits)) = split_len.last_key_value() {
+                // let split_internal_nodes = biggest_splits.iter().map(|(n, split_for)|); // TODO select by internal node count?
```
Comment (birchmd) [inline lib/relooper/src/graph/supergraph.rs]: Please remove commented out code if it is not needed.

File: lib/relooper/src/traversal/graph/dfs.rs [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+pub fn dfs_post<T, ChIt, ChFun>(start: T, get_children: &mut ChFun) -> Vec<T>
+where
+    T: Hash + Eq + Copy,
+    ChIt: IntoIterator<Item = T>,
+    ChFun: FnMut(&T) -> ChIt,
+{
+    let mut visited: HashSet<T> = HashSet::from([start]);
+    let mut res: Vec<T> = Vec::new();
+
+    dfs_post_inner(start, get_children, &mut res, &mut visited); //TODO rewrite using Iterator or at least without recursion
```
Comment (birchmd) [inline lib/relooper/src/traversal/graph/dfs.rs]: Remaining todo

File: lib/relooper/src/traversal/graph/dfs.rs [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+pub fn dfs_post_ord<T, ChIt, ChFun>(start: T, get_children: &mut ChFun) -> Vec<T>
+where
+    T: Ord + Eq + Copy,
+    ChIt: IntoIterator<Item = T>,
+    ChFun: FnMut(&T) -> ChIt,
+{
+    let mut visited: BTreeSet<T> = BTreeSet::from([start]);
+    let mut res: Vec<T> = Vec::new();
+
+    dfs_post_inner_ord(start, get_children, &mut res, &mut visited); //TODO rewrite using Iterator or at least without recursion
```
Comment (birchmd) [inline lib/relooper/src/traversal/graph/dfs.rs]: Remaining todo.

File: tools/apply-functions.py [hunk 25]
```diff
@@ -0,0 +1,16 @@
+functions = {}
```
Comment (birchmd) [inline tools/apply-functions.py]: Unless there is some documentation on how to use this script, I don't think we should commit it to the public repo.

File: tools/compile-contract.py [hunk 26]
```diff
@@ -0,0 +1,31 @@
+import sys
```
Comment (birchmd) [inline tools/compile-contract.py]: Similarly here I don't think it makes sense to commit to the public repo unless there is some documentation on how to use it.

File: tools/int-to-hex.clj [hunk 27]
```diff
@@ -0,0 +1,19 @@
+;; converts all int/long literals to hex
```
Comment (birchmd) [inline tools/int-to-hex.clj]: Even if there was documentation for this one, I still wouldn't recommend committing it to the public repo. I don't think we want to expand the number of languages a user needs to have installed on their system to use the tooling. It could at least be written in Python to be consistent with the other scripts.

**Review by birchmd — COMMENTED** [review]
I'm pretty happy with this overall! Most of my comments are about code quality (adding comments, resolving todos).

One other note: it's nice for your reviewers if you include a description of the PR. You can focus the reviewer's attention to specific parts of the PR that are especially important. And it's helpful for our future selves as well if we come back to this PR in the future and wonder what the heck it was trying to do.

File: bin/evm2near/src/compile.rs (hunk 5 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: Tests shows that it's work, but @shishkin-pavel can you explain it more detailly

File: bin/evm2near/src/compile.rs (hunk 9 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: I think this case really can be marked as unreachable. Also, I tried to add ``unreachable!()``  and code with it successfully compiled all 5 solidity contracts. @shishkin-pavel WDYT ?

File: bin/evm2near/src/compile.rs (hunk 5 above)
Comment (shishkin-pavel) [reply bin/evm2near/src/compile.rs]: unlike `_start`function in C-like languages (runtime initialization routine), `_start` in wasm seems to be just a default entry point
so, wasmtime invocation without entry point specification will start that function and return "succesfully", which is probably not the thing we meant to do anyways, thats why I thought that getting process terminated with error is better than silently do nothing and pass

File: lib/evmlib/src/ops.rs (hunk 12 above)
Comment (MCJOHN974) [reply lib/evmlib/src/ops.rs]: I think it's still necessary to correctly burned gas counting. But we really don't call it anywhere. Maybe now we have an incorrect gas counting? @shishkin-pavel

File: bin/evm2near/src/compile.rs (hunk 7 above)
Comment (shishkin-pavel) [reply bin/evm2near/src/compile.rs:320]: there are some thoughts:
* every syntactical nesting in initial source code can produce one nesting level in that structure
* some optimizations should be able to reduce nesting (but we are not in control of them)
* im doubting, there is optimizations adding nesting levels (because nesting represents control flow, and control flow is relatively expensive usually, so we need to reduce it)
* to deal with irreducible graphs we need to copy initial cfg parts, so it can produce additional nesting
* as a joke: https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1061 

in conclusion: it definitely can produce stack overflow and should be rewritten, but probably we can deal with that later (and chances are high that stack overflow will never happen on any human-written code)

File: bin/evm2near/src/compile.rs (hunk 9 above)
Comment (shishkin-pavel) [reply bin/evm2near/src/compile.rs]: we can reach that point only in case if we are falling out of exit node without `revert`/`return`/... opcode, which, I believe, should never happen in correct evm bytecode (but Im not sure)
feels right to me to mark it as `unreachable!` and investigate failing cases (if any)

File: bin/evm2near/src/compile.rs (hunk 5 above)
Comment (birchmd) [reply bin/evm2near/src/compile.rs]: According to the [wasm specification](https://www.w3.org/TR/wasm-core-1/#start-function%E2%91%A0)

> The [start](https://www.w3.org/TR/wasm-core-1/#syntax-module) component of a module declares the [function index](https://www.w3.org/TR/wasm-core-1/#syntax-funcidx) of a start function that is automatically invoked when the module is [instantiated](https://www.w3.org/TR/wasm-core-1/#exec-instantiation), after [tables](https://www.w3.org/TR/wasm-core-1/#syntax-table) and [memories](https://www.w3.org/TR/wasm-core-1/#syntax-mem) have been initialized.

And you can see [this happening](https://docs.rs/wasmer-vm-near/2.4.0/src/wasmer_vm/instance/mod.rs.html#958-972) in the Wasm library the Near Runtime uses.

Since running contracts works without hitting the `unreachable` I think it must mean we are not actually assigning `_start` as being the official start function of the wasm module. Perhaps this is something to look into.

File: lib/evmlib/src/ops.rs (hunk 12 above)
Comment (shishkin-pavel) [reply lib/evmlib/src/ops.rs]: actually no, these functions will never be called, its safe to delete them

File: bin/evm2near/src/compile.rs (hunk 5 above)
Comment (shishkin-pavel) [reply bin/evm2near/src/compile.rs]: oh, thanks! its quite strange that wasmtime doesnt call that function either

File: bin/evm2near/src/analyze.rs (hunk 1 above)
Comment (MCJOHN974) [reply bin/evm2near/src/analyze.rs:14]: I think it's fixed

File: bin/evm2near/src/analyze.rs (hunk 2 above)
Comment (MCJOHN974) [reply bin/evm2near/src/analyze.rs:26]: I think it's fixed

File: bin/evm2near/src/analyze.rs (hunk 3 above)
Comment (MCJOHN974) [reply bin/evm2near/src/analyze.rs]: I think it's fixed

File: bin/evm2near/src/analyze.rs (hunk 4 above)
Comment (MCJOHN974) [reply bin/evm2near/src/analyze.rs]: I think it's fixed

File: bin/evm2near/src/compile.rs (hunk 8 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: I think it's fixed

File: bin/evm2near/src/compile.rs (hunk 9 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: I think it's fixed

File: bin/evm2near/src/compile.rs (hunk 10 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: I think it's fixed

File: lib/relooper/BACKLOG.md (hunk 13 above)
Comment (MCJOHN974) [reply lib/relooper/BACKLOG.md]: I think it's fixed

File: lib/relooper/README.md (hunk 14 above)
Comment (MCJOHN974) [reply lib/relooper/README.md]: I think it's fixed

File: lib/relooper/src/graph/supergraph.rs (hunk 22 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/supergraph.rs]: I think it's fixed

File: tools/apply-functions.py (hunk 25 above)
Comment (MCJOHN974) [reply tools/apply-functions.py]: I think it's fixed

File: tools/compile-contract.py (hunk 26 above)
Comment (MCJOHN974) [reply tools/compile-contract.py]: I think it's fixed

File: tools/int-to-hex.clj (hunk 27 above)
Comment (MCJOHN974) [reply tools/int-to-hex.clj]: I think it's fixed

File: lib/relooper/src/graph/supergraph.rs (hunk 21 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/supergraph.rs]: I think it's fixed

File: lib/evmlib/src/api.rs (hunk 11 above)
Comment (shishkin-pavel) [reply lib/evmlib/src/api.rs:153]: my suggestion would be to create an issue for gas-accounting-related things instead of fixing that small thing now

Comment (birchmd) [reply lib/evmlib/src/api.rs:153]: Sure, that's fine with me.

File: bin/evm2near/src/compile.rs (hunk 7 above)
Comment (birchmd) [reply bin/evm2near/src/compile.rs:320]: Cool, thanks for the explanation. Sounds like it is not an immediate concern. Maybe just file as a GitHub issue that we can come back to later.

File: bin/evm2near/src/analyze.rs (hunk 2 above)
Comment (birchmd) [reply bin/evm2near/src/analyze.rs:26]: Personally I don't think "serial number" is very clear. Maybe adding an example would help. Something like for the "EVM bytecode `0x60806040` the offset for the second opcode (`0x80`) is `Off(1)` and the serial number is `Idx(???)`".

File: lib/relooper/src/graph/supergraph.rs (hunk 22 above)
Comment (birchmd) [reply lib/relooper/src/graph/supergraph.rs]: Thanks for deleting the commented out code. But the todo remains, will something be done about it or should it be turned into a GitHub issue?

File: bin/evm2near/src/compile.rs (hunk 5 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: Fixed

File: bin/evm2near/src/analyze.rs (hunk 2 above)
Comment (MCJOHN974) [reply bin/evm2near/src/analyze.rs:26]:
```
0x00  PUSH
0x03  PUSH
0x06  ADD
```
Here first ``PUSH`` will have ``idx = 0`` and ``offs = 0x00``, second ``idx = 1`` and ``offs = 0x03``, ``ADD`` will have ``idx = 2`` and ``offs = 0x06``. Both this numbers are used in ``compile.rs``

File: lib/relooper/src/graph/supergraph.rs (hunk 22 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/supergraph.rs]: Hm, I think this function contains quite big and interesting task that can be an issue. 

There is a lot of reducible graphs that are equivalent to given irreducible, but they are different in the size. We want generate new graph as small as possible. In the paper about relooper was given a greedy algorithm, Pavel implemented some more complicated algorithm. But we don't have proofs if that algorithm is best possible, and I think it is not the best. So, I think some time in the future it will be good to make research about this algorithms and implement the best one.

File: bin/evm2near/src/analyze.rs (hunk 2 above)
Comment (birchmd) [reply bin/evm2near/src/analyze.rs:26]: Great, that looks like a nice example to add to the doc comment in the code.

Comment (MCJOHN974) [reply bin/evm2near/src/analyze.rs:26]: I added it

File: lib/relooper/src/graph/cfg/mod.rs (hunk 17 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: I think it's fixed

**Comment by MCJOHN974** [comment]
Also I add unit tests to CI in [this commit](https://github.com/aurora-is-near/evm2near/pull/9/commits/e0882bbbc7b40a044216f9b3f0b2701d12919b8e)

File: lib/relooper/src/graph/mod.rs (hunk 19 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/mod.rs]: @shishkin-pavel is it any sense to put declaration of ``EnrichedCfg`` and some implementations of some it's functions to ``mod.rs`` instead of ``enrichments.rs`` ?

Comment (MCJOHN974) [reply lib/relooper/src/graph/mod.rs]: I think fixed in [this commit](https://github.com/aurora-is-near/evm2near/pull/9/commits/c4064be82494c1ecfd3b6233f4d0cb65e9dccdde)

File: lib/relooper/src/graph/cfg/mod.rs (hunk 16 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: I think there is no way to avoid this clone. You need insert ``n`` to ``self.nodes`` and later, in line 103, insert ``edge`` into ``self.outedges``.  And both this addings are moving

Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: @shishkin-pavel

File: bin/evm2near/src/compile.rs (hunk 6 above)
Comment (MCJOHN974) [reply bin/evm2near/src/compile.rs]: I think it's fixed [here](https://github.com/aurora-is-near/evm2near/pull/9/commits/63bb464fa7e1330f907250c618a03a44835f92eb)

File: lib/relooper/src/graph/cfg/mod.rs (hunk 15 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: Unfortunately it's is not trivial problem with borrow checker, I think now is better just left it as is

File: lib/relooper/src/graph/cfg/mod.rs (hunk 16 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: Unfortunately it's is not trivial problem with borrow checker, I think now is better just left it as is

File: lib/relooper/src/graph/cfg/mod.rs (hunk 18 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: Unfortunately it's is not trivial problem with borrow checker, I think now is better just left it as is

File: lib/relooper/src/graph/cfg/mod.rs (hunk 15 above)
Comment (birchmd) [reply lib/relooper/src/graph/cfg/mod.rs]: Some help: https://github.com/MCJOHN974/evm2near/pull/12

File: lib/relooper/src/graph/cfg/mod.rs (hunk 18 above)
Comment (birchmd) [reply lib/relooper/src/graph/cfg/mod.rs]: I suggest rewriting the function as follows:

```rust
    pub fn strip_unreachable(&mut self) {
        let unreachable_nodes: HashSet<TLabel> = self
            .nodes()
            .difference(&self.reachable_nodes())
            .into_iter()
            .map(|n| **n)
            .collect();
        for unreachable in unreachable_nodes {
            self.out_edges.remove(&unreachable);
        }
    }
```

The reason we cannot avoid doing at least one copy is because we need to use the same `out_edges` collection for all the operations here. If we do not copy the `TLabel` data out of the collection then we cannot remove it (removing from the collection would invalidate the pointers we created while traversing over `out_edges` in the first place). But in the way I have refactored it we minimize how many nodes need to be copy. We only copy after taking the difference instead of copying all the nodes twice.

File: lib/relooper/src/graph/cfg/mod.rs (hunk 15 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: Thanks!!

File: lib/relooper/src/graph/cfg/mod.rs (hunk 18 above)
Comment (MCJOHN974) [reply lib/relooper/src/graph/cfg/mod.rs]: Thanks! I rewrite it [here](https://github.com/aurora-is-near/evm2near/pull/9/commits/c8c5aa752c29308f4ccd130834ebb8b88f979377)

Comment (birchmd) [reply lib/relooper/src/graph/cfg/mod.rs]: Thanks @MCJOHN974 . The change was accidentally reverted in [this commit](https://github.com/aurora-is-near/evm2near/pull/9/commits/0c07fd79731e7e3b7bfd86bfaff07fc2c64a8200). Please push the rewrite again.

File: lib/evmlib/src/api.rs (hunk 11 above)
Comment (MCJOHN974) [reply lib/evmlib/src/api.rs:153]: Fixed [here](https://github.com/aurora-is-near/evm2near/pull/9/commits/1a6999de1851f64f2d35a1dd972c47cc40175ba3)

File: lib/relooper/src/traversal/graph/dfs.rs (hunk 23 above)
Comment (MCJOHN974) [reply lib/relooper/src/traversal/graph/dfs.rs]: Fixed [here](https://github.com/aurora-is-near/evm2near/pull/9/commits/1a6999de1851f64f2d35a1dd972c47cc40175ba3)

File: lib/relooper/src/traversal/graph/dfs.rs (hunk 24 above)
Comment (MCJOHN974) [reply lib/relooper/src/traversal/graph/dfs.rs]: Fixed [here](https://github.com/aurora-is-near/evm2near/pull/9/commits/1a6999de1851f64f2d35a1dd972c47cc40175ba3)

**Comment by MCJOHN974** [comment]
@birchmd we fixed all your notes or decided to make from them new big tasks. Please, take a look

**Review by artob — APPROVED** [review]
Great work, @shishkin-pavel and @MCJOHN974!
